channel2/wss_listener.go

```diff
@@ -0,0 +1,280 @@
+}
+
+func (c *WssConnection) Close() error {
+	return c.socket.Close()
+}
+
+var upgrader = websocket.Upgrader{} // use default options
+
+/**
+ *	Accept incoming HTTP connection, and upgrade it to a websocket suitable for comms between browZer and Ziti edge router
+ */
+func (listener *wssListener) handleWebsocket(w http.ResponseWriter, r *http.Request) {
+	log := pfxlog.ContextLogger(listener.endpoint.String())
+	log.Info("entered")
+
+	upgrader.CheckOrigin = func(r *http.Request) bool { return true } // Allow all origins
+
+	c, err := upgrader.Upgrade(w, r, nil) // upgrade from HTTP to binary socket
+
+	if err != nil {
+		log.WithField("err", err).Error("websocket upgrade failed. Failure not recoverable. Exiting listen loop")
+	} else {
+
+		socket := c.UnderlyingConn() // Obtain the socket underneath the websocket
+
+		connection := &WssConnection{
+			detail: &transport.ConnectionDetail{
+				Address: "wss:" + socket.RemoteAddr().String(),
+				InBound: true,
+				Name:    "wss",
+			},
+			websocket: c,
+			socket:    socket.(*tls.Conn),
+		}
+
+		listener.incoming <- connection // pass the socket to the goroutine that will validate the HELLO handshake
+	}
+}
+
+/**
+ *	Simply start the goroutines that will process incoming websocket connections
+ */
+func (listener *wssListener) Listen(handlers ...ConnectionHandler) error {
+	go listener.wsslistener()
+	go listener.binaryListener()
+	return nil
+}
+
+/**
+ *	The TLS-based listener that accepts incoming HTTP connections that we need to upgrade to websocket connections
+ */
+func (listener *wssListener) wsslistener() {
+	log := pfxlog.ContextLogger(listener.endpoint.String())
+	log.Info("starting HTTP (websocket) server")
+
+	router := mux.NewRouter()
+
+	router.HandleFunc("/wss", listener.handleWebsocket).Methods("GET")
+
+	httpServer := &http.Server{
+		Addr:         listener.endpoint.BindableAddress(),
+		WriteTimeout: time.Second * 15,
+		ReadTimeout:  time.Second * 15,
+		IdleTimeout:  time.Second * 60,
+		Handler:      router,
+		TLSConfig: &tls.Config{
+			ClientAuth: tls.RequestClientCert,
+		},
+	}
+
+	// log.Info("listener.options: [%v]", listener.options)
+
+	if err := httpServer.ListenAndServeTLS(listener.serverCert, listener.key); err != nil {
+		panic(err)
+	}
+}
+
+func (listener *wssListener) Close() error {
+	close(listener.close)
+	close(listener.created)
+	if err := listener.socket.Close(); err != nil {
+		return err
+	}
+	listener.socket = nil
+	return nil
+}
+
+func (listener *wssListener) Create(tcfg transport.Configuration) (Underlay, error) {
+	if listener.created == nil {
+		return nil, errors.New("closed")
+	}
+	impl := <-listener.created
+	if impl == nil {
+		return nil, errors.New("closed")
+	}
+	return impl, nil
+}
+
+/**
+ *	Take the incoming (now binary) websocket
+ */
+func (listener *wssListener) binaryListener() {
+	log := pfxlog.ContextLogger(listener.endpoint.String())
+	log.Info("started")
+	defer log.Info("exited")
```

Comment:
This type of logging needs to

1) be `Debug()`or `Trace()`
2) have more context than "exited"

You are probably using this as working-level information and we don't have module-level logging flags so I get it. Just don't leave it in.

Reply:
these were temp log msgs, and have been removed